ArjunaJTA/jta/tests/classes/com/hp/mwtests/ts/jta/recovery/RecoveryMonitorTestDeadlock.java

```diff
@@ -0,0 +1,133 @@
+ * in the <code>SUSPENDED</code> state.
+ * The shutdown request waits for the recovery listeners to finish but the recovery listeners are waiting for the periodic recovery
+ * <code>wakeUp</code> call to complete which never happens because periodic recovery is suspended.
+ */
+public class RecoveryMonitorTestDeadlock {
+    private static Logger log = Logger.getLogger(RecoveryMonitorTestDeadlock.class.getName());
+    private static final int TEST_TIMEOUT_S = 1;
+
+    @Test
+    public void testRecoveryMonitorWithSuccess() throws Exception {
+        final CountDownLatch countdown = new CountDownLatch(1);
+        // XAResource which suspends the recover call to be able to run suspend on recovery manager meanwhile
+        final XAResource sleepingXA = new XATestResource(XATestResource.OK_JNDI_NAME, false) {
+            @Override
+            public Xid[] recover(int i) throws XAException {
+                countdown.countDown();
+                try {
+                    log.fine(this + ": Recovering for " + TEST_TIMEOUT_S + " seconds");
+                    TimeUnit.SECONDS.sleep(TEST_TIMEOUT_S);
+                } catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
+                }
+                System.out.println(this + ": Recovered.");
+                return new Xid[0];
+            }
+        };
+        // resource helper to be registered to recovery manager for it to process the XAResource
+        final XAResourceRecoveryHelper xaResourceRecoveryHelper = new XAResourceRecoveryHelper() {
+            @Override
+            public boolean initialise(String p) throws Exception {
+                return true;
+            }
+            @Override
+            public XAResource[] getXAResources() throws Exception {
+                return new XAResource[] {sleepingXA};
+            }
+        };
+
+        RecoveryEnvironmentBean recoveryEnvironmentBean = recoveryPropertyManager.getRecoveryEnvironmentBean();
+        recoveryEnvironmentBean.setRecoveryBackoffPeriod(1); // use a short interval between passes
+        recoveryEnvironmentBean.setRecoveryListener(true); // configure the RecoveryMonitor
+
+        RecoveryManager manager = RecoveryManager.manager(RecoveryManager.DIRECT_MANAGEMENT);
+        XARecoveryModule recoveryModule = new XARecoveryModule();
+        recoveryModule.addXAResourceRecoveryHelper(xaResourceRecoveryHelper);
+        manager.addModule(recoveryModule); // we only need to test the XARecoveryModule
+        manager.startRecoveryManagerThread(); // start periodic recovery
+
+        String host = recoveryEnvironmentBean.getRecoveryAddress(); // the recovery listener host
+        String rcPort = String.valueOf(recoveryEnvironmentBean.getRecoveryPort()); // the recovery listener port
+
+        // running the recovery scan which will be delayed by sleepingXAResource
+        manager.scan(null);
+        countdown.await(TEST_TIMEOUT_S, TimeUnit.SECONDS);
```

Comment:
So you trigger a background scan and then wait for 1 second for the latch to hit zero
(and PeriodicRecovery will take at least one second because the XAR#recovery call sleeps for 1 second).
I would expect that the countdown.await call would timeout rather than hit zero in this case. Can you add an assert here to indicate what your expectation is.

Reply:
yes, pretty good point, thanks. I will change this.